File: frontend/types/task.ts
/**
 * Mirrors backend/app/schemas/task.py - keep these in sync manually
 * (same convention as types/planner.ts, types/goal.ts).
 */

export type PlanTaskStatus = "pending" | "completed";

export interface PlanTask {
  id: string;
  plan_id: string;
  milestone_id: string | null;
  week_number: number;
  day_number: number | null;
  title: string;
  description: string | null;
  estimated_minutes: number | null;
  display_order: number;
  status: PlanTaskStatus;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface WeekTasksGroup {
  week_number: number;
  tasks: PlanTask[];
}

export interface PlanTasksResponse {
  weeks: WeekTasksGroup[];
}

export interface Progress {
  completed: number;
  total: number;
  remaining: number;
  percentage: number;
}

File: backend/app/schemas/task.py
"""Request/response schemas for the Execution Engine (PlanTask + progress)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.plan_task import PlanTaskStatus


class PlanTaskRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    plan_id: uuid.UUID
    milestone_id: uuid.UUID | None
    week_number: int
    day_number: int | None
    title: str
    description: str | None
    estimated_minutes: int | None
    display_order: int
    status: PlanTaskStatus
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class WeekTasksGroup(BaseModel):
    week_number: int
    tasks: list[PlanTaskRead]


class PlanTasksResponse(BaseModel):
    weeks: list[WeekTasksGroup]


class UpdateTaskStatusRequest(BaseModel):
    """Pydantic's enum validation is the whole guard here - anything
    other than "pending"/"completed" (see PlanTaskStatus) is rejected
    with a 422 before this ever reaches task_service, matching the
    "status should initially support pending/completed" constraint."""

    status: PlanTaskStatus


class ProgressRead(BaseModel):
    completed: int
    total: int
    remaining: int
    percentage: int
